app/controllers/application_controller.rb

```diff
@@ -97,11 +97,36 @@ def mnemosyne_trace
     end
   end
 
-  def switch_locale(&) # rubocop:disable Metrics/AbcSize
-    session[:locale] = sanitize_locale(params[:custom_locale] || params[:locale] || session[:locale])
-    locale = session[:locale] || http_accept_language.compatible_language_from(I18n.available_locales) || I18n.default_locale
+  def sanitized_locale_param
+    sanitize_locale(params[:locale])
+  end
+
+  def sanitized_session_locale
+    sanitize_locale(session[:locale])
+  end
+
+  def users_preferred_locale
+    return if current_user.nil?
+
+    sanitize_locale(current_user.preferred_locale)
+  end
+
+  def choose_locale
+    (sanitized_locale_param ||
+      sanitized_session_locale ||
+      users_preferred_locale ||
+      http_accept_language.compatible_language_from(I18n.available_locales) ||
+      I18n.default_locale).to_sym
+  end
+
+  def set_locale
+    locale = choose_locale
+    I18n.locale = locale
+    session[:locale] = locale
     Sentry.set_extras(locale:)
-    I18n.with_locale(locale, &)
+    if current_user.present? && locale != users_preferred_locale
```

Comment:
Yes, I will fix that.

Reply:
Ah, right, because sanitising a locale will also convert it to a symbol. I missed that. However, in that case, the additional call `to_sym` in `choose_locale` is unnecessary, isn't it? The first three (`sanitized...`) will always return a symbol or `nil`. The fourth option will return a value of the list we pass as argument, and the list only contains of symbols. Similarly, the last option also is a symbol.

Reply:
That's true. I will remove that.